File: ruby/luhn/luhn.rb
module Luhn
  def self.valid?(str)
    str = str.gsub(/\s/, '')

    return false if str =~ /\D/
    return false if str.size < 2

    digits = str.chars.map(&:to_i).reverse

    sum = digits.each.with_index.sum do |n, i|
      if i.even?
        n
      else
        n *= 2

        if n > 9
          n - 9
        else
          n
        end
      end
    end

    return sum % 10 == 0
  end
end

File: javascript/luhn/luhn.js
export const valid = (str) => {
    str = str.replace(/\s/g, "");

    if (/\D/.test(str))
        return false;

    if (str.length < 2)
        return false;

    let digits = Array.from(str).map(x => parseInt(x)).reverse();

    let sum = digits.reduce(function(acc, val, idx) {
        if (idx % 2 == 0) {
            return acc + val;
        } else {
            val *= 2;

            if (val > 9)
                return acc + val - 9;
            else
                return acc + val;
        }
    });

    return sum % 10 == 0;
};
